### ISSUE TITLE
When HYPHEN-MINUS is on the extremity of a word, it should not be turned into HYPHEN

### ISSUE THREAD

**Issue by jesus2099** [kellnerd/musicbrainz-bookmarklets#20]
Hello @kellnerd,

It happens to me that I have to manually _type_ EN DASH instead of HYPHEN-MINUS that have been turned into HYPHEN.
It must be difficult to handle all cases.

I think a HYPHEN-MINUS should be replaced to HYPHEN only when it is surrounded by `\S`.
Otherwise it's used like parenthesis and should be a DASH (EN DASH more than EM DASH, because of its length proximity with HYPHEN-MINUS): https://musicbrainz.org/edit/86608589 (manual)

But that's not that easy.
With some CJK tracklists, there may not always be spaces around those DASH that are used like parenthesis.
But we cannot guess everything.

**Comment by kellnerd** [comment]
It was actually intended to have a leading or trailing HYPHEN-MINUS converted into HYPHEN because these are very common in German (e.g. let's say you have two compound words "AX & BX" and are to lazy to spell out X two times, then you can simply write "A- & BX" &ndash; or "XA & -B" instead of "XA & XB"). But I don't know if this concept also exists in other languages, I think most of them rather use spaces instead of hyphens to build compound words ("X A & B" instead of "X A & X B").

I think it should be easy to add an additional rule for dashes which are used as brackets, it would look very similar to the existing rules for quoted text. I am just not sure if there is a need for it, I have never encountered these in a tracklist so far. Shouldn't we standardise these to real parentheses for ETI per the style guidelines anyway? Or is there an exception for CJK tracklists? (Sorry, but I have no experience in this area 😅)

**Comment by jesus2099** [comment]
Ah yes indeed, we keep them for the [Japanese releases](https://wiki.musicbrainz.org/Style/Language/Japanese) that I edit a lot (and they use quite a few other characters than parentheses), but yes I remember outside of Japanese releases, they are standardised to parentheses, indeed!

It didn't know the German HYPHEN pattern and didn't see it in the 4 languages I know, at least. :)

I don't know, we could test the release language: `document.querySelector("select#language").value == 198 /* Japanese, but is this ID stable? */` which also work when we are in the Tracklist tab.

Or maybe limit the German HYPHEN pattern on German releases (`145`)?
Are these language ID stable?

But it makes the thing more and more comlpex. ;)

**Comment by kellnerd** [comment]
Thanks for the information and the idea, a language-dependent rule could indeed be the solution.
I already have a function which replaces the quoting rules for certain languages (based on the release language), it would be easy to prepend* an additional rule for Japanese in there:

https://github.com/kellnerd/musicbrainz-bookmarklets/blob/74c98bf27c6e10a2f99b03cae79baf8e820c11ea/src/guessUnicodePunctuation.js#L83

Maybe something like this would work (but it would have no effect for your example release with English transliteration):

```js
if (language === 'Japanese') {
	rules.unshift([/(?<=[^\p{L}\d]|^)-(.+?)-(?=[^\p{L}\d]|$)/ug, '–$1–'])
}
```

> Are these language ID stable?

I hope so, AFAIK languages do not have MBIDs and [I'm already using these internal numbers](https://github.com/kellnerd/musicbrainz-bookmarklets/blob/main/src/languages.js).

*) The support for the German HYPHEN pattern is more or less a bonus effect of the lazy implementation of the (current) last rule, which simply replaces all the remaining occurrences of HYPHEN-MINUS with HYPHEN:

https://github.com/kellnerd/musicbrainz-bookmarklets/blob/74c98bf27c6e10a2f99b03cae79baf8e820c11ea/src/guessUnicodePunctuation.js#L36

But I should probably make that rule stricter, i.e. use `/\S-\S/g` as you suggested. Then the German rules could be extended the same way as I outlined above for Japanese.

**Comment by kellnerd** [comment]
I had already made the hyphen rule stricter in ea1e32aa07f318466398304eb5ccc1f76b626e7b, before I added support for additional language-specific punctuation rules in f15debf3be8508b0fb3fe4199842137422ce33e2 (released in v2022.1.28.2).
The example that I used to test the rule for "dashes which are used as hyphens" probably isn't the best (as it isn't really a Japanese title), so suggestions for better test cases are welcomed.
